### PR TITLE
RFD: `tctl sso configure` command.

### PR DESCRIPTION
**Pull request by Tener**
This RFD is "implement auth connector configurator" part of #9270.

### CODE REVIEW

File: rfd/0049-sso-configure-command.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+Each combination of `--kind` and `--provider` will be mapped to single template based
+on [text/template](https://pkg.go.dev/text/template) package. The extension points will be handled by calls to functions
+or conditionals, with functions handling the input fetching and state management.
+
+## Security
+
+The command will never modify existing Teleport configuration, this will be done by the user by further invocation
+of `tctl create` command or by other means.
+
+The command may handle user secrets. The implementation must ensure these are silently written anywhere (e.g. we must
```
Comment (r0mant) [inline rfd/0049-sso-configure-command.md:96]: Did you mean to say "these are **not** silently written anywhere"?

File: rfd/0049-sso-configure-command.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+Valid flag combinations (initial implementation scope):
+
+| `--kind` | `--preset` | interactive behavior                                      | batch behaviour                    |
+|----------|------------|-----------------------------------------------------------|------------------------------------|
+|          |            | Ask for `--kind` to use.                                  | Exit with error: missing `--kind`. |
+| `github` |            | Ask for `client_id` and `client_secret`.                  |                                    |
+| `oidc`   |            | Ask for `--preset`.                                       | Default to `--preset=generic`.     |
+| `oidc`   | `generic`  | Generic OIDC template.                                    |                                    |
```
[the remark this answers is not in the document]
Comment (Tener) [reply rfd/0049-sso-configure-command.md:48]: I'll simplify and drop --kind entirely, so `--kind=oidc --preset=generic` will now be simply `--preset=oidc`.

File: rfd/0049-sso-configure-command.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+```bash
+$ tctl sso configure
+```
+
+Users provide choices using command line flags. The `--kind` flag determines the kind of auth connector,
+while `--preset`
+chooses template suited for particular vendor:
+
+```bash
+$ tctl sso configure --kind={oidc,saml,github} --preset={google,okta,adfs,generic} --interactive --template-specific-flags=...
```
[the remark this answers is not in the document]
Comment (Tener) [reply rfd/0049-sso-configure-command.md:38]: Sounds good.

File: rfd/0049-sso-configure-command.md (hunk 1 above)
Comment (Tener) [reply rfd/0049-sso-configure-command.md:96]: Yep, good catch.

File: rfd/0049-sso-configure-command.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+## Proxy URL discovery
+
+The command will make an effort to discover the correct proxy URL. Failing that, an appropriate default along with
+comment will be put into the configuration file.
+
+## Implementation details
+
+Each combination of `--kind` and `--provider` will be mapped to single template based
+on [text/template](https://pkg.go.dev/text/template) package. The extension points will be handled by calls to functions
```
[the remark this answers is not in the document]
Comment (Tener) [reply rfd/0049-sso-configure-command.md:88]: Sure, this makes things simpler if less flexible (we cannot output ill-formed YAML with those methods, so there is no way of adding comments and similar stuff). But this may be a good thing after all: the implementation will be simpler and more testable.

File: rfd/0049-sso-configure-command.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+We support two modes: interactive and batch (non-interactive). In interactive mode we may provide the user with
+choices (e.g. preset) or ask them to provide details (secrets, identifiers etc.) In batch mode we never ask but use
+defaults (when possible) or insert comments asking for the data to be filled manually afterwards. Some
+
+Interactive mode is used when we detect TTY or `--interactive` flag is provided, batch mode is used otherwise.
+
+## Proxy URL discovery
+
+The command will make an effort to discover the correct proxy URL. Failing that, an appropriate default along with
```
[the remark this answers is not in the document]
Comment (Tener) [reply rfd/0049-sso-configure-command.md:82]: Ok.

File: rfd/0049-sso-configure-command.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+Detailed help message can be access with:
+
+```bash
+$ tctl sso configure --help
+...
+```
+
+## Interactive and batch modes
+
+We support two modes: interactive and batch (non-interactive). In interactive mode we may provide the user with
```
[the remark this answers is not in the document]
Comment (Tener) [reply rfd/0049-sso-configure-command.md:74]: I rephrased this a bit.

File: rfd/0049-sso-configure-command.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+choices (e.g. preset) or ask them to provide details (secrets, identifiers etc.) In batch mode we never ask but use
+defaults (when possible) or insert comments asking for the data to be filled manually afterwards. Some
```
[the remark this answers is not in the document]
Comment (Tener) [reply rfd/0049-sso-configure-command.md:76]: Thanks, this was a leftover from a deleted sentence.

**Comment by Tener** [comment]
@xinding33 @r0mant @smallinsky

Please take a look at this updated version.

File: rfd/0062-tctl-sso-configure-command.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+- assertion encryption (support varies by IdP)
+- externally-provided signing key (by default Teleport will self-issue this)
+- overrides for specific IdP providers
+
+```
+--signing_key_file         A file with request signing key. Must be used together with --signing_cert_file.
+--signing_cert_file        A file with request certificate. Must be used together with --signing_key_file.
+--assertion_key_file       A file with key used for securing SAML assertions. Must be used together with --assertion_cert_file.
+--assertion_cert_file      A file with cert used for securing SAML assertions. Must be used together with --assertion_key_file.
```
Comment (r0mant) [inline rfd/0062-tctl-sso-configure-command.md:92]: I'm pretty sure that our CLI flags use dashes instead of underscores so I'd rename all these new flags to match. I left the same comment in your implementation PR.

**Review by r0mant — APPROVED** [review]
lgtm with the comment about flags.

**Comment by r0mant** [comment]
@xinding33 Can you give this another quick look from the UX perspective and approve if it looks good so we can merge it? This basically describes the same `tctl sso configure` command we saw on a demo a couple weeks ago.

File: rfd/0062-tctl-sso-configure-command.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
+---
+authors: Krzysztof Skrzętnicki <krzysztof.skrzetnicki@goteleport.com>
+state: draft
+---
+
+# RFD 62 - `tctl sso configure` command
```
Comment (r0mant) [inline rfd/0062-tctl-sso-configure-command.md:6]: Sorry but I stole number 62 in https://github.com/gravitational/teleport/blob/master/rfd/0062-tsh-proxy-template.md so don't forget to update this :)

File: rfd/0062-tctl-sso-configure-command.md (hunk 8 above)
Comment (Tener) [reply rfd/0062-tctl-sso-configure-command.md]: Thanks, done.

File: rfd/00XX-tctl-sso-configure-command.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+### Flags: SAML
+
+For `SAML` the commonly used flags will be:
+
+```
+-p, --preset                   Preset. One of: [okta onelogin ad adfs]
+-n, --name                     Connector name. Required, unless implied from preset.
+-e, --entity-descriptor        Set the Entity Descriptor. Valid values: file, URL, XML content. Supplies configuration parameters as single XML instead of individual elements.
+-a, --attributes-to-roles      Sets attribute-to-role mapping in the form 'attr_name,attr_value,role1,role2,...'. Repeatable.
```
Comment (ben-mase) [inline rfd/00XX-tctl-sso-configure-command.md:58]: Looking at the examples, I find this syntax a bit hard to parse without going back and referencing the docs. I know it's tricky because the user is having to specify three or more things at once. There might not be a better solution, but I found myself wondering about alternative syntaxes like `-a group[developer]=access`

Comment (Tener) [reply rfd/00XX-tctl-sso-configure-command.md:58]: Unfortunately, both `[` and `]` are special characters in many shells which makes them poor choices. But I'm happy to consider alternative syntaxes that do not have this problem.

**Review by smallinsky — APPROVED** [review]
